Finished translation of "Passing Props to a Component" page (#495)
reactjs/uk.react.dev: src/content/learn/passing-props-to-a-component.md

---
title: Передача пропсів до компонента
---

<Intro>

Компоненти React використовують *пропси* для взаємодії між собою. Кожен батьківський компонент може передати деяку інформацію своїм дочірнім компонентам, передаючи їм пропси. Пропси можуть нагадувати вам HTML-атрибути, але ви можете передавати через них будь-яке значення JavaScript, включаючи об'єкти, масиви та функції.

</Intro>

<YouWillLearn>

* Як передавати пропси до компонента
* Як читати пропси з компонента
* Як встановлювати значення за замовчуванням для пропсів
* Як передавати JSX до компонента
* Як пропси змінюються з часом

</YouWillLearn>

## Знайомі пропси {/*familiar-props*/}

> Пропси — це інформація, яку ви передаєте до тегу JSX. Наприклад, `className`, `src`, `alt`, `width` та `height` деякі з пропсів, які ви можете передати до тегу `<img>`:

Reviewer: ```suggestion
Пропси — це інформація, яку ви передаєте до тегу JSX. Наприклад, `className`, `src`, `alt`, `width` та `height` — деякі з пропсів, які ви можете передати до тегу `<img>`:
```

<Note>

Якщо подвійні фігурні дужки після `person=` вас збивають з пантелику, згадайте, що [вони просто об'єкт](/learn/javascript-in-jsx-with-curly-braces#using-double-curlies-css-and-other-objects-in-jsx) всередині фігурних дужок JSX.

</Note>

Тепер ви можете читати ці пропси всередині компонента `Avatar`.

### Крок 2: Читайте пропси всередині дочірнього компонента {/*step-2-read-props-inside-the-child-component*/}

Ви можете прочитати ці пропси, перераховуючи їх назви `person, size`, розділені комами всередині `({` та `})` безпосередньо після `function Avatar`. Це дозволяє використовувати їх всередині коду `Avatar` як змінні.

```js
function Avatar({ person, size }) {
  // person та size доступні тут
}
```

> Додайте деяку логіку до `Avatar`, яка використовує пропси `person` та `size` для відображення і готово.

Reviewer: ```suggestion
Додайте трохи логіки до `Avatar`, яка використовує пропси `person` та `size` для відображення, і готово.
```

> Тепер ви можете налаштувати `Avatar` для відображення багатьма різними способами з різними пропсами. Спробуйте змінити значення!

Reviewer: Note: here `render` is translated as `відображення`, though it's `рендер` elsewhere and in the translation guide.

Author: Yes, you're right, it's my mistake 😅

I will correct it

</Sandpack>

> Props дозволяють вам думати про батьківські та дочірні компоненти незалежно. Наприклад, ви можете змінити проп `person` або `size` всередині `Profile`, не думаючи про те, як `Avatar` використовує їх. Аналогічно, ви можете змінити спосіб використання `Avatar` цих пропсів, не звертаючи уваги на `Profile`.

Reviewer: Props or пропси

> Ви можете уявляти пропси як "ручки", які ви можете налаштовувати. Вони виконують ту саму роль, що і аргументи для функцій - насправді, пропси _є_ єдиним аргументом вашого компонента! Функції компонентів React приймають один аргумент - об'єкт `props`:

Reviewer: ```suggestion
Уявіть пропси як «ручки регулювання», які ви можете налаштовувати. Вони виконують ту саму роль, що і аргументи для функцій — насправді, пропси _є_ єдиним аргументом вашого компонента! Функції компонентів React приймають один аргумент — об'єкт `props`:
```

}
```

Зазвичай вам не потрібен весь об'єкт `props` сам по собі, тому ви розкладаєте його на окремі пропси.

<Pitfall>

**Не пропустіть пару фігурних дужок `{` та `}` всередині `(` та `)` при оголошенні пропсів:

```js
function Avatar({ person, size }) {
  // ...
}
```

> Цей синтаксис називається ["деструктуризація"](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Unpacking_fields_from_objects_passed_as_a_function_parameter) і еквівалентний читанню властивостей з параметра функції:

Reviewer: ```suggestion
Цей синтаксис називається ["деструктуризація"](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Unpacking_fields_from_objects_passed_as_a_function_parameter) і є еквівалентним читанню властивостей із параметра функції:
```

Author: ```suggestion
Цей синтаксис називається ["деструктуризацією"](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Unpacking_fields_from_objects_passed_as_a_function_parameter) і еквівалентний витягуванню властивостей з параметра функції:
```

What do you think of this option? It's a bit paraphrased, but in my opinion, it's more understandable

Reviewer: @ltlaitoff it's up to you :) I don't mind personal approach to translation

}
```

> Це передає всі пропси `Profile` до `Avatar` без перерахування кожного з їх імен.

Reviewer: Note: above `name` was translated as `назв` and not `імен`. we should get consistent there.

Author: Yes, you're right, it's my mistake 😅

</Sandpack>

Спробуйте замінити `<Avatar>` всередині `<Card>` на деякий текст, щоб побачити, як компонент `Card` може обгорнути будь-який вкладений вміст. Він не потребує "знати", що саме рендериться всередині нього. Ви ще побачите цей гнучкий шаблон в багатьох місцях.

> Ви можете уявляти компонент з пропом `children` як компонент з "отвіром", який може бути "заповнений" батьківськими компонентами з довільним JSX. Ви часто будете використовувати властивість `children` для візуальних обгорток: панелей, сіток і т.д.

Author: > btw, I believe, we should have «these quotation marks» everywhere instead of "these ones".

To do this, we need to add it to the translation rules, and as far as I understand, in other already translated articles, we use "text" specifically. That's why I believe, for now, we should continue using it.

@bkatsevych WDYT?

> Ви можете уявляти компонент з пропом `children` як компонент з "отвіром", який може бути "заповнений" батьківськими компонентами з довільним JSX. Ви часто будете використовувати властивість `children` для візуальних обгорток: панелей, сіток і т.д.

Reviewer: As for now there's a problem that both variants of quotation marks are used

<Illustration src="/images/docs/illustrations/i_children-prop.png" alt='Подібна до головоломки плитка Card з слотами для "дочірніх" шматочків, наприклад текст або Avatar' />

## Як пропси змінюються з часом {/*how-props-change-over-time*/}

Компонент `Clock`, наведений нижче, отримує два пропи від батьківського компонента: `color` та `time`. (Код батьківського компонента опущено, оскільки він використовує [стан](/learn/state-a-components-memory), про який ми ще не говорили.)

> Спробуйте змінити колір у випадаючому списку нижче:

Author: I think the upper option will be better

@bkatsevych WDYT?

Однак, пропси є [незмінними(immutable)](https://uk.wikipedia.org/wiki/%D0%9D%D0%B5%D0%B7%D0%BC%D1%96%D0%BD%D0%BD%D0%B8%D0%B9_%D0%BE%D0%B1%27%D1%94%D0%BA%D1%82) — термін з комп'ютерних наук, що означає "незмінні". Коли компоненту потрібно змінити свої пропси (наприклад, відповідно до взаємодії користувача або нових даних), він повинен "запитати" свій батьківський компонент передати йому _інші пропси_ — новий об'єкт! Його старі пропси будуть відкинуті, і, врешті-решт, рушій JavaScript звільнить пам'ять, яку вони займали.

**Не намагайтеся «змінювати пропси»**. Коли вам потрібно реагувати на введення користувача (наприклад, змінювати вибраний колір), вам потрібно «встановити стан», про що ви можете дізнатися в [Стан: Пам'ять компонента.](/learn/state-a-components-memory)

<Recap>

* Щоб передати пропси, додайте їх до JSX, так само, як ви робите з атрибутами HTML.
* Щоб прочитати пропси, використовуйте синтаксис деструктуризації `function Avatar({ person, size })`.
* Ви можете вказати значення за замовчуванням, наприклад, `size = 100`, яке використовується для відсутніх та `undefined` пропсів.

> * Ви можете передати всі пропси JSX за допомогою синтаксису розширення `<Avatar {...props} />`, але не зловживайте цим!

Author: I took this translation from the book 'uk javascript info'

In your opinion, what would be the most appropriate translation for this term? And I suggest adding it to the dictionary

> * Ви можете передати всі пропси JSX за допомогою синтаксису розширення `<Avatar {...props} />`, але не зловживайте цим!

Reviewer: @ltlaitoff my point was solely about using either `spread` itself or its translation.

But yeah, I definitely would like to have a good translation and have it added to our translation guide.
Maybe, one of the best variant is  `поширення/поширити` (vue docs translators used it), as `розширення` is mostly associated with add-on/extension.

@bkatsevych wdyt?

</Sandpack>

Хоча синтаксис виглядає трохи інакше, оскільки ви описуєте властивості об'єкта JavaScript, а не колекцію атрибутів JSX, ці приклади в основному еквівалентні, і ви можете вибрати будь-який підхід.

</Solution>

#### Налаштування розміру зображення на основі пропу {/*adjust-the-image-size-based-on-a-prop*/}

У цьому прикладі компонент `Avatar` отримує числовий проп `size`, який визначає ширину та висоту елемента `<img>`. Для пропу `size` задано значення `40`. Однак, якщо ви відкриєте зображення в новій вкладці, ви помітите, що саме зображення є більшим (`160` пікселів). Справжній розмір зображення визначається розміром ескізу, який ви запитуєте.

> Змініть компонент `Avatar`, щоб він запитував найближчий розмір зображення на основі пропу `size`. Зокрема, якщо `size` менше `90`, передайте `'s'` (від англ. «small» — малий) замість `'b'` (від англ. «big» — великий) у функцію `getImageUrl`. Переконайтеся, що ваші зміни працюють, виконуючи рендер аватарів з різними значеннями пропу `size` та відкриваючи зображення в новій вкладці.

Reviewer 2: ```suggestion
Змініть компонент `Avatar`, щоб він запитував найближчий розмір зображення на основі пропу `size`. Зокрема, якщо `size` менше `90`, передайте `'s'` (від англ. _«small»_ — малий) замість `'b'` (від англ. _«big»_ — великий) у функцію `getImageUrl`. Переконайтеся, що ваші зміни працюють, виконуючи рендер аватарів з різними значеннями пропу `size` та відкриваючи зображення в новій вкладці.
```
